csa.swift

```diff
@@ -0,0 +1,125 @@
+import Foundation
+
+typealias Station = Int
+
+protocol InputType {
+    func read(count: Int) throws -> [Int]?
+}
+
+struct Error: ErrorType {
+    let message: String
+    init(_ message: String) {
+        self.message = message
+    }
+}
+
+struct Connection {
+    let from: Station, to: Station, departure: time_t, arrival: time_t
+    
+    init?(input: InputType) throws {
+        guard let values = try input.read(4) else { return nil }
+        from = values[0]
+        to = values[1]
+        departure = values[2]
+        arrival = values[3]
+        guard from >= 0 && to >= 0 else { throw Error("Station number cannot be negative.") }
+        guard arrival >= departure else { throw Error("Time travel is not supported yet.") }
+    }
+}
+
+struct Request {
+    let from: Station, to: Station, departure: time_t
+
+    init?(input: InputType) throws {
+        guard let values = try input.read(3) else { return nil }
+        from = values[0]
+        to = values[1]
+        departure = values[2]
+        guard from >= 0 && to >= 0 else { throw Error("Station number cannot be negative.") }
+    }
+}
+
+struct Router {
+    private let connections: [Connection]
+    private let count: Int
+    
+    init(input: InputType) throws {
+        var connections: [Connection] = []
+        while let connection = try Connection(input: input) {
+            connections.append(connection)
+        }
+        self.connections = connections
+        self.count = connections.reduce(0) { (count, connection) in
+            max(count, connection.from + 1, connection.to + 1)
+        }
+    }
+    
+    func solve(request: Request) -> [Connection] {
+        let from = request.from, to = request.to, departure = request.departure
+        guard count > max(from, to) else { return [] }
+        var incoming: [Connection?] = Array(count: count, repeatedValue: nil)
+        for connection in connections {
```

Comment:
You’ve reversed all the conditions from the reference implementations 😁.

I kinda like the pattern `if let <thing> where <condition on thing>`, it’s somewhat self-explaining.

On the other hand the loop itself has C-style imperative logic (`continue`,`break`) and I feel those Swift-y patterns don’t really help in that case. What’s your opinion on this?